File: android/src/main/kotlin/com/navideck/universal_ble/UniversalBleLogger.kt
package com.navideck.universal_ble

import android.util.Log
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

object UniversalBleLogger {
    private const val TAG = "UniversalBle"
    private var currentLogLevel: BleLogLevel = BleLogLevel.NONE
    private val timeFormatter = SimpleDateFormat("HH:mm:ss.SSS", Locale.US)

    fun setLogLevel(logLevel: BleLogLevel) {
        currentLogLevel = logLevel
    }

    fun logError(message: String) {
        if (!allows(BleLogLevel.ERROR)) return
        Log.e(TAG, withTimestamp(message))
    }

    fun logWarning(message: String) {
        if (!allows(BleLogLevel.WARNING)) return
        Log.w(TAG, withTimestamp(message))
    }

    fun logInfo(message: String) {
        if (!allows(BleLogLevel.INFO)) return
        Log.i(TAG, withTimestamp(message))
    }

    fun logDebug(message: String) {
        if (!allows(BleLogLevel.DEBUG)) return
        Log.d(TAG, withTimestamp(message))
    }

    fun logVerbose(message: String) {
        if (!allows(BleLogLevel.VERBOSE)) return
        Log.v(TAG, withTimestamp(message))
    }

    private fun allows(level: BleLogLevel): Boolean {
        return currentLogLevel != BleLogLevel.NONE && level.ordinal <= currentLogLevel.ordinal
    }

    private fun withTimestamp(message: String): String {
        return "[${timeFormatter.format(Date())}] $message"
    }
}

File: darwin/universal_ble/Sources/universal_ble/UniversalBleLogger.swift
import Foundation

final class UniversalBleLogger {
  static let shared = UniversalBleLogger()

  private init() {}

  private lazy var dateFormatter: DateFormatter = {
    let formatter = DateFormatter()
    formatter.dateFormat = "HH:mm:ss.SSS"
    return formatter
  }()

  private var currentLogLevel: UniversalBleLogLevel = .none

  func setLogLevel(_ logLevel: UniversalBleLogLevel) {
    currentLogLevel = logLevel
  }

  func logError(_ message: String) {
    guard allows(.error) else { return }
    print("UniversalBle:ERROR \(withTimestamp(message))")
  }

  func logWarning(_ message: String) {
    guard allows(.warning) else { return }
    print("UniversalBle:WARN \(withTimestamp(message))")
  }

  func logInfo(_ message: String) {
    guard allows(.info) else { return }
    print("UniversalBle:INFO \(withTimestamp(message))")
  }

  func logDebug(_ message: String) {
    guard allows(.debug) else { return }
    print("UniversalBle:DEBUG \(withTimestamp(message))")
  }

  func logVerbose(_ message: String) {
    guard allows(.verbose) else { return }
    print("UniversalBle:VERBOSE \(withTimestamp(message))")
  }

  private func allows(_ level: UniversalBleLogLevel) -> Bool {
    return currentLogLevel != .none && level.rawValue <= currentLogLevel.rawValue
  }

  private func withTimestamp(_ message: String) -> String {
    let time = dateFormatter.string(from: Date())
    return "[\(time)] \(message)"
  }
}
